### PR TITLE
support multiple particles in Arcs API

### PR DESCRIPTION
**Pull request by mmandlis**

### CODE REVIEW

**Comment by mmandlis** [comment]
I realize this is not the most elegant way to achieve this, but it is the fastest and simplest until we figure out what we want - kotlin, parcelables, autovalue, protos etc, whatever idk...

File: javaharness/java/arcs/android/api/IArcsService.aidl [hunk 1]
```diff
@@ -15,9 +16,9 @@ interface IArcsService {
       String arcId,
       String pecId,
       String recipe,
-      String particleId,
-      String particleName,
-      String providedSlotId,
+      in List<String> particleIds,
+      in List<String> particleNames,
+      in List<String> providedSlotIds,
```
Comment (csilvestrini) [inline javaharness/java/arcs/android/api/IArcsService.aidl]: I don't really love having parallel lists here. This would be better as a `List<ParticleData>` or something like that, but it seems a bit cumbersome to do. I found https://developer.android.com/guide/components/aidl#PassingObjects which suggests you could make ParticleData implement Parcelable. But the ParticleData.particle member will cause problems there...

File: javaharness/java/arcs/android/client/ArcsServiceBridge.java [hunk 2]
```diff
@@ -42,16 +42,25 @@
...
-                callback));
+    runServiceMethod(service -> {
+      List<String> particleIds = new ArrayList<>();
+      List<String> particleNames = new ArrayList<>();
+      List<String> providedSlots = new ArrayList<>();
+      arcData.getParticleList().forEach(particleData -> {
+        particleIds.add(particleData.getId());
+        particleNames.add(particleData.getName());
+        providedSlots.add(particleData.getProvidedSlotId());
+      });
```
Comment (csilvestrini) [inline javaharness/java/arcs/android/client/ArcsServiceBridge.java]: This would also be easier if it accepted a List<ParticleData>

File: javaharness/java/arcs/android/demo/service/ArcsAutofillService.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public void onFillRequest(
     ArcData arcData = arcs.runArc("AndroidAutofill", autofillParticle);
 
     autofillRenderer.addCallback(
-        arcData.getProvidedSlotId(),
+        arcData.getParticleList().get(0).getProvidedSlotId(),
```
Comment (csilvestrini) [inline javaharness/java/arcs/android/demo/service/ArcsAutofillService.java]: Add a TODO?

File: javaharness/java/arcs/android/impl/ArcsLocal.java [hunk 4]
```diff
@@ -34,10 +34,16 @@
 
   @Override
   public void runArc(ArcData arcData) {
-    if (arcData.getParticle() != null) {
-      PECInnerPort pecInnerPort =
-          pecPortManager.getOrCreateInnerPort(arcData.getPecId(), arcData.getSessionId());
-      pecInnerPort.mapParticle(arcData.getParticle());
+    PECInnerPort pecInnerPort = null;
+    for (int i = 0; i < arcData.getParticleList().size(); ++i) {
```
Comment (csilvestrini) [inline javaharness/java/arcs/android/impl/ArcsLocal.java]: for (ArcData.ParticleData particleData : arcData.getParticleList()) { ... }

File: javaharness/java/arcs/android/api/IArcsService.aidl (hunk 1 above)
Comment (mmandlis) [reply javaharness/java/arcs/android/api/IArcsService.aidl]: yeah, this is where i started, and I think ideally the entire ArcData would be Parcelable (I think they can be nested, so ParticleData would be Parcelable too).
But then i talked to Ray @cromwellian and he proposed I used AutoValue. But I also have logic in this class. 
And if/when we switch to Kotlin, according to Ray we get the parcelable for free, but then probably need a separate class with the logic vs the one in the API, not sure..

Bottom line, it's not completely clear to me what is the right approach (and also this API will be deprecated once we migrate to use the storage API), and this isn't the highest priority task, so I don't have too much time to spend on it, but OTOH i want to get it done.

So I was hoping to get away with this PR as an incremental progress and a TODO.

File: javaharness/java/arcs/android/impl/ArcsLocal.java (hunk 4 above)
Comment (mmandlis) [reply javaharness/java/arcs/android/impl/ArcsLocal.java]: done

File: javaharness/java/arcs/android/api/IArcsService.aidl (hunk 1 above)
Comment (csilvestrini) [reply javaharness/java/arcs/android/api/IArcsService.aidl]: Fair enough. I guess I'm just generally concerned about the number of hacks and TODOs we've been adding, without actually addressing the underlying code health issues.

What you have here would be ok for now, but I think we should settle on a proper solution like AutoValue/Kotlin/protos sooner rather than later..

Comment (mmandlis) [reply javaharness/java/arcs/android/api/IArcsService.aidl]: i agree. 
but i would also like to mention that we did do some refactoring and resolved some of the technical debt after the prototype was done. this is not to say everything is awesome. this is to say we aren't completely ignoring it...
